working/2365 - primary constructors/feature-specification.md

```diff
@@ -0,0 +1,472 @@
+declaration is an `inline` class or an `enum` declaration, the modifier
+`final` on every instance variable is required. Hence, it can be omitted:
+
+```dart
+inline class I {
+  final int x;
+  I.name(this.x);
+}
+
+inline class I.name(int x);
+
+class Point {
+  final int x;
+  final int y;
+  const Point(this.x, this.y);
+}
+
+class const Point(int x, int y);
+
+enum E {
+  one('a'),
+  two('b');
+
+  final String s;
+  const E(this.s);
+}
+
+enum E(String s) { one('a'), two('b') }
+```
+
+This mechanism follows an existing pattern, where `const` modifiers can be
+omitted in the case where the immediaty syntactic context implies that this
+modifier _must_ be present. For example `const [const C()]` can be written
+as `const [C()]`. In the examples above, the parameter-and-variable
+declarations `final int x` and `final int y` are written as `int x` and
+`int y`, and this is allowed because it would be a compile-time error to
+omit `final` in an `inline` class, and in a class with a constant
+constructor. In other words, when we see `inline` on the class or `const`
+on the class name, we know that `final` is implied on all instance
+variables.
+
+A proposal which was mentioned during the discussions about primary
+constructors was that the keyword `final` could be used in order to specify
+that all instance variables introduced by the primary constructor are
+`final` (but the constructor wouldn't be constant, and hence there's more
+freedom in the declaration of the rest of the class). However, that
+proposal is not included here, because it may be a source of confusion that
+`final` may also occur as a modifier on the class itself, and also because
+the resulting class header does not contain syntax which is already similar
+to a non-primary constructor declaration.
+`class final Point(int x, int y);` cannot use the similarity to a
+non-primary constructor declaration to justify the keyword `final`.
+
+```dart
+class Point {
+  final int x;
+  final int y;
+  Point(this.x, this.y);
+}
+
+class final Point(int x, int y); // Not supported!
+```
+
+Optional parameters can be declared as usual in a primary constructor, with
+default values that must be constant as usual:
+
+```dart
+class Point {
+  int x;
+  int y;
+  Point(this.x, [this.y = 0]);
+}
+
+class Point(int x, [int y = 0]);
+```
+
+Similarly for named parameters, required or not:
+
+```dart
+class Point {
+  int x;
+  int y;
+  Point(this.x, {required this.y});
+}
+
+class Point(int x, {required int y});
+```
+
+The class header can have additional elements, just like class headers
+where there is no primary constructor:
+
+```dart
+class D<TypeVariable extends Bound> extends A with M implements B, C {
+  final int x;
+  final int y;
+  const D.named(this.x, [this.y = 0]);
+}
+
+class const D.named<TypeVariable extends Bound>(int x, [int y = 0])
+    extends A with M implements B, C;
+```
+
+There was a proposal from Bob that the primary constructor should be
```

Comment:
This is getting quite discussion based, and long (we're still in the Introduction?)

Reply:
Done.